text/0056-tx-fee-map-digest.md

```diff
@@ -0,0 +1,95 @@
+
+To protect from this, we need to ensure that the fee map the client is using to come up with the fee they are paying is the same one that is configured on the node. We propose adding a new field to the `Tx` structure that contains the digest of the fee map. This allows the enclave to verify that the fee map the client is using matches its own, and if not to reject the transaction without revealing any information about it.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We will modify the `Tx` structure to look like this:
+```
+pub struct Tx {
+    /// The transaction contents.
+    #[prost(message, required, tag = "1")]
+    pub prefix: TxPrefix,
+
+    /// The transaction signature.
+    #[prost(message, required, tag = "2")]
+    pub signature: SignatureRctBulletproofs,
+
+    /// Client's belief about the minimum fee map, expressed as a merlin digest.
+    ///
+    /// The enclave must reject the proposal if this doesn't match the enclave's
+    /// belief, to protect the client from information disclosure attacks.
+    /// (This is TOB-MCCT-5)
+    #[prost(bytes, tag = "3")]
+    pub fee_map_digest: Vec<u8>,
+}
+```
+
+Clients wanting to protect themselves from this attack will begin including the minimum fee map digest inside the `Tx`. We will introduce a new method to `TransactionBuilder`, `set_fee_map` that takes a `FeeMap` and uses it to populate `fee_map_digest` when the transaction is constructured. `FeeMap` will need to be modified to include a method to calculate a digest, so that all users of `FeeMap` calculate the digest in the same way.
+We will also need to move `FeeMap` into a different crate, since right now it is inside `mc-consensus-enclave-config`. However, following the changes proposed here, `TransactionBuilder` will also need to import the `FeeMap` object so that it can calculate its digest when constructing a transaction (this might also be useful for having sane default minimum fees). It doesn't make sense that `TransactionBuilder` will have a depdency on any consensus enclave crates.
+
+Since this is a new field, and it can be empty, it is entirely backwards compatible. Old clients will not include it, and in that case the enclave will not attempt to compare it to its own `FeeMap` digest. It will only get included by clients that wish to protect themselves from this attack.
+
+Clients can learn about the FeeMap a node is configured with by using the `BlockchainAPI.GetLastBlockInfo` GRPC API. While there is no guarantee the node isn't lying about the fee map it hands the user, the fact that the enclave will later compare the `fee_map_digest` the client computes from the fee map handed by the node to the one it has been initialized with provide the protection from the attack described above.
```

Comment:
we might want to write something about how clients should handle this error

i think if they try to refresh their value for fee map, and ensuring they get a value from several nodes, not just the one they got the error from, then it should be enough to mitigate this